### PR TITLE
fix: Add `maxWait` to `sendUpdate` debounce to fix UI sync starvation

### PR DESCRIPTION
**Pull request by MajorLift**
## **Description**

Adds `{ maxWait: MILLISECOND * 1000 }` to the `debounce` options for `sendUpdate` in `metamask-controller.js`, forcing a periodic flush even under continuous state updates. This prevents the background-to-UI state sync from being starved indefinitely during heavy operations (e.g. syncing >100 accounts).

[Open in GitHub Codespaces](https://codespaces.new/MetaMask/metamask-extension/pull/40331?quickstart=1)

## **Changelog**

CHANGELOG entry: Fixed background-to-UI state sync starvation that caused UI to become stuck indefinitely during large account syncs.

## **Related issues**

Fixes: https://github.com/MetaMask/MetaMask-planning/issues/6777

## **Manual testing steps**

1. Load an SRP with >100 accounts
2. Open the Accounts list
3. Verify accounts populate and the UI does not get stuck in a permanent "syncing" state

## **Pre-merge author checklist**

- [x] I've followed [MetaMask Contributor Docs](https://github.com/MetaMask/contributor-docs) and [MetaMask Extension Coding Standards](https://github.com/MetaMask/metamask-extension/blob/main/.github/guidelines/CODING_GUIDELINES.md).
- [x] I've completed the PR template to the best of my ability
- [x] I've included tests if applicable
- [x] I've documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [x] I've applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-extension/blob/main/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

---

> [!NOTE]
> **Medium Risk**
> Adjusts background-to-UI state update scheduling; behavioral/timing changes could affect UI freshness and event ordering under heavy state churn. Includes an e2e test change to reduce flakiness, but edge-case timing regressions are still possible.
> 
> **Overview**
> Prevents background-to-UI state sync starvation by adding `maxWait: SECOND` to the `sendUpdate` debounce in `metamask-controller.js`, ensuring state updates flush periodically even during continuous updates.
> 
> Hardens MetaMetrics user-trait generation in `metametrics-controller.ts` by making several `preferences`-derived traits null-safe with defaults, and updates the metrics e2e test to tolerate identify events being split across multiple updates by polling/merging traits before asserting.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit c512fb67269cfae34c8e991742e491f02eb8ed66. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

### CODE REVIEW

File: app/scripts/metamask-controller.js [hunk 1]
```diff
@@ -492,6 +492,7 @@ export default class MetamaskController extends EventEmitter {
     this.sendUpdate = debounce(
       this.privateSendUpdate.bind(this),
       MILLISECOND * 200,
+      { maxWait: MILLISECOND * 1000 }, // Force flush to avoid indefinite sync starvation
```
[the remark this answers is not in the document]
Comment (MajorLift) [reply app/scripts/metamask-controller.js]:
```suggestion
      { maxWait: SECOND }, // Force flush to avoid indefinite sync starvation
```

File: app/scripts/controllers/metametrics-controller.ts [hunk 2]
```diff
@@ -1441,7 +1441,7 @@ export default class MetaMetricsController extends BaseController<
       [MetaMetricsUserTrait.TokenDetectionEnabled]:
         metamaskState.useTokenDetection,
       [MetaMetricsUserTrait.ShowNativeTokenAsMainBalance]:
-        metamaskState.preferences.showNativeTokenAsMainBalance,
+        metamaskState.preferences?.showNativeTokenAsMainBalance,
```
Comment (MajorLift) [inline app/scripts/controllers/metametrics-controller.ts]:
```suggestion
        metamaskState.preferences?.showNativeTokenAsMainBalance ?? false,
```

File: app/scripts/controllers/metametrics-controller.ts [hunk 3]
```diff
@@ -1452,11 +1452,11 @@ export default class MetaMetricsController extends BaseController<
       [MetaMetricsUserTrait.HasMarketingConsent]:
         metamaskState.dataCollectionForMarketing,
       [MetaMetricsUserTrait.TokenSortPreference]:
-        metamaskState.preferences.tokenSortConfig?.key || '',
+        metamaskState.preferences?.tokenSortConfig?.key || '',
       [MetaMetricsUserTrait.PrivacyModeEnabled]:
-        metamaskState.preferences.privacyMode,
+        metamaskState.preferences?.privacyMode,
```
Comment (MajorLift) [inline app/scripts/controllers/metametrics-controller.ts]:
```suggestion
        metamaskState.preferences?.privacyMode ?? false,
```

**Review by HowardBraham — APPROVED** [review]

File: app/scripts/metamask-controller.js [hunk 4]
```diff
@@ -492,6 +492,7 @@ export default class MetamaskController extends EventEmitter {
     this.sendUpdate = debounce(
       this.privateSendUpdate.bind(this),
       MILLISECOND * 200,
+      { maxWait: SECOND }, // Force flush to avoid indefinite sync starvation
```
Comment (MajorLift) [inline app/scripts/metamask-controller.js:495]: Core change

File: app/scripts/controllers/metametrics-controller.ts [hunk 5]
```diff
@@ -1441,7 +1441,7 @@ export default class MetaMetricsController extends BaseController<
       [MetaMetricsUserTrait.TokenDetectionEnabled]:
         metamaskState.useTokenDetection,
       [MetaMetricsUserTrait.ShowNativeTokenAsMainBalance]:
-        metamaskState.preferences.showNativeTokenAsMainBalance,
+        metamaskState.preferences?.showNativeTokenAsMainBalance ?? false,
```
Comment (MajorLift) [inline app/scripts/controllers/metametrics-controller.ts:1444]: The nullish fallback values in this file are based on [`getDefaultPreferencesControllerState`](https://github.com/MetaMask/metamask-extension/blob/b4ceb11be202b9204e61e47461bf24797df4c149/app/scripts/controllers/preferences-controller.ts#L183-L210)

File: test/e2e/tests/metrics/segment-user-traits.spec.ts [hunk 6]
```diff
@@ -165,10 +165,38 @@ describe('Segment User Traits', function () {
...
         await privacySettings.toggleDataCollectionForMarketing();
-        events = await getEventPayloads(driver, mockedEndpoints);
-        assert.equal(events.length, 1);
-        assert.deepStrictEqual(events[0].traits.is_metrics_opted_in, true);
-        assert.deepStrictEqual(events[0].traits.has_marketing_consent, true);
+        // maxWait on sendUpdate debounce may split the two toggles into
+        // separate identify events. Poll until the expected traits arrive.
+        await driver.wait(async () => {
+          try {
+            events = await getEventPayloads(driver, mockedEndpoints, false);
```
Comment (MajorLift) [inline test/e2e/tests/metrics/segment-user-traits.spec.ts:172]: - Added polling logic for metametrics segment traits test. 
- Handles edge case where toggle state updates are not batched together due to `maxWait`.

**Review by MajorLift — COMMENTED** [review]
Notes for reviewers:

**Review by hjetpoluru — APPROVED** [review]
